src/GitServer/GitServer.ts

```diff
@@ -0,0 +1,298 @@
+import fs from 'fs'
+import url from 'url'
+import Path from 'path'
+import http from 'http'
+import { promisify } from "util";
+import { AddressInfo } from "net";
+import { Readable } from "stream";
+import { parse } from 'querystring'
+import { HttpDuplex } from "./HttpDuplex";
+import uploadPack from "./upload-pack/UploadPack";
+import VaultStore from "@polykey/VaultStore/VaultStore";
+
+// Here is the protocol git outlines for sending pack files over http:
+// https://git-scm.com/docs/pack-protocol/2.17.0
+// https://github.com/git/git/blob/master/Documentation/technical/pack-protocol.txt
+// This should be consulted in developing our upload pack implementation
+
+// This git backend (as well as HttpDuplex class) is heavily inspired by node-git-server:
+// https://github.com/gabrielcsapo/node-git-server
+
+const services = ['upload-pack', 'receive-pack']
+
+class GitServer {
+  repoDir: string;
+  vaultStore: VaultStore;
+  autoCreate = false
+  server: http.Server;
+  constructor(
+    repoDir: string,
+    vaultStore: VaultStore
+  ) {
+    this.repoDir = repoDir
+    this.vaultStore = vaultStore
+  }
+
+  /**
+   * Find out whether `repoName` exists.
+   */
+  exists(repo: string) {
+    // TODO: consider if vault has been shared
+    return fs.existsSync(Path.join(this.repoDir, repo))
+  }
+
+  /**
+   * returns the typeof service being process
+   */
+  getType(service: any): string {
+    switch (service) {
+      case 'upload-pack':
+        return 'fetch'
+      case 'receive-pack':
+        return 'push'
+      default:
+        return 'unknown'
+    }
+  }
+
+
+  // alternatively you can also pass authenticate a promise
+  authenticate(
+    type: string,
+    repo: string,
+    username: string,
+    password: string,
+    headers: http.IncomingHttpHeaders
+  ): boolean {
+    if(username === 'foo') {
+      return true
+    }
+    return false
+  }
+
+  /**
+   * Handle incoming HTTP requests with a connect-style middleware
+   */
+  handle(req: http.IncomingMessage, res: http.ServerResponse) {
+
+    const infoHandler = (req: http.IncomingMessage, res: http.ServerResponse): boolean => {
```

Comment:
Perhaps these handlers should be abstracted in the MVC style like loja-api? Not sure if you copied this structure from node-git-server?

Reply:
yeah this structure is directly from `node-git-sever`. I will try to simplify this along with the HttpDuplex class

Reply:
I'm not sure what you mean by the MVC style from loja-api, these are just for handling different GET or POST requests. I think I can understand where you're going though. Perhaps it's better if we have a single handler that determines what type of action is needed (GET/POST) and then have different methods defined for handling the IncomingMessage?

Reply:
This has now been simplified and I was able to remove quite a bit of code. I think its a little easier to read now :+1: